controller/src/main/java/org/jboss/as/controller/ModelControllerLock.java

```diff
@@ -102,12 +143,31 @@ protected synchronized boolean tryAcquire(int permit) {
         }
 
         @Override
-        protected synchronized boolean tryRelease(int permit) {
-            final Object value = permitHolder.get();
-            if (value == null) {
+        protected int tryAcquireShared(final int permit) {
+            if (permitHolder != null) {
+                return -1;
+            }
+            if (compareAndSetState(0, 1)) {
+                return 1;
+            } else {
```

Comment:
Yeah, this has been bugging me too, I've been looking at this code quite a bit today. Internally I think the read / write could synchronize on permitholder, or a bitmask in the internal state perhaps. That would at least allow just the single getValue() and CAS in those methods, which is how it seems to be intended to work.